python/pyspark/sql/functions.py

```diff
@@ -3262,7 +3262,19 @@ def regexp_extract(str: "ColumnOrName", pattern: str, idx: int) -> Column:
     return _invoke_function("regexp_extract", _to_java_column(str), pattern, idx)
 
 
-def regexp_replace(str: "ColumnOrName", pattern: str, replacement: str) -> Column:
+@overload
+def regexp_replace(string: "ColumnOrName", pattern: str, replacement: str) -> Column:
+    ...
+
+
+@overload
+def regexp_replace(string: "ColumnOrName", pattern: Column, replacement: Column) -> Column:
+    ...
+
+
+def regexp_replace(
+    string: "ColumnOrName", pattern: Union[str, Column], replacement: Union[str, Column]
```

Comment:
Can we write up `Parameters` section in the docs? And I think you can only keep `pattern: Union[str, Column]` and remove the ones above.

Reply:
Added Parameters to the docstring: [45cd79d](https://github.com/apache/spark/pull/37329/commits/45cd79ddd205ff317f0061fdcf19972f84bbee91)

The types make it somewhat unclear that pattern and replacement must be either _both_ strings or _both_ columns to match the signatures [here](https://github.com/apache/spark/blob/775b5ced70c87d6a8709f44eb4a5de48c286a51d/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L2826). This is why I had the overloaded signatures. I removed them and tried to clarify in the docstring instead.

Reply:
> The types make it somewhat unclear that pattern and replacement must be either both strings or both columns to match the signatures

Yeah, I would go with `overload` here as well. We capture these types of dependencies in other places and I don't see any reason to make type checks weaker here.

Reply:
Technically speaking we could also default to invoking `(Column, Column) -> Column` and convert arguments accordingly, i.e.

```python
     return _invoke_function(
            "regexp_replace",
            _to_java_column(string),
            _create_column_from_literal(pattern) if isinstance(pattern, str) else _to_java_column(pattern),
            _create_column_from_literal(replacement) if isinstance(replacement, str) else _to_java_column(replacement),
        )
```

Reply:
Great ideas. I implemented these in [https://github.com/apache/spark/pull/37329/commits/0bdd92bd07b4d5661049f594bf34d0595ab019c4](0bdd92b). Black formatting pushed the if-else statements onto three lines, which I thought was less readable. I changed the formatting a bit to help with readability.